### PR TITLE
GUVNOR-3547: [Guided Decision Table] Patterns from BRL conditions are not available for another action columns

### PR DESCRIPTION
**Pull request by manstis**
See https://issues.jboss.org/browse/GUVNOR-3547

### CODE REVIEW

**Comment by manstis** [comment]
Part of an ensemble:

- https://github.com/kiegroup/drools/pull/1529
- https://github.com/kiegroup/drools-wb/pull/648

File: drools-wb-screens/drools-wb-guided-dtable-editor/drools-wb-guided-dtable-editor-client/src/main/java/org/drools/workbench/screens/guided/dtable/client/widget/table/GuidedDecisionTablePresenter.java [hunk 1]
```diff
@@ -761,18 +762,13 @@ public boolean hasColumnDefinitions() {
 
     @Override
     public boolean canConditionBeDeleted(final ConditionCol52 col) {
```
Comment (manstis) [inline drools-wb-screens/drools-wb-guided-dtable-editor/drools-wb-guided-dtable-editor-client/src/main/java/org/drools/workbench/screens/guided/dtable/client/widget/table/GuidedDecisionTablePresenter.java]: I noticed deletion of ```BRLConditionColumn``` was incorrectly being handled by ```canConditionBeDeleted( final ConditionCol52 col )``` (as it is a sub-class of it).. so fixed to use the correct check for the applicable class/sub-class.

File: drools-wb-screens/drools-wb-guided-dtable-editor/drools-wb-guided-dtable-editor-client/src/main/java/org/drools/workbench/screens/guided/dtable/client/wizard/column/commons/HasPatternPage.java [hunk 2]
```diff
@@ -30,5 +30,5 @@
 
     String getEntryPointName();
 
-    List<PatternWrapper> getPatterns();
+    Set<PatternWrapper> getPatterns();
```
Comment (manstis) [inline drools-wb-screens/drools-wb-guided-dtable-editor/drools-wb-guided-dtable-editor-client/src/main/java/org/drools/workbench/screens/guided/dtable/client/wizard/column/commons/HasPatternPage.java]: Changed from a ```List``` to ```Set``` (as the old code contained duplicate entries for multiple actions updating the same pattern)... this has a knock-on effect across a handful of other classes.

File: drools-wb-screens/drools-wb-guided-dtable-editor/drools-wb-guided-dtable-editor-client/src/main/java/org/drools/workbench/screens/guided/dtable/client/wizard/column/plugins/ActionSetFactPlugin.java [hunk 3]
```diff
@@ -222,27 +224,32 @@ public void setEntryPointName(final String entryPointName) {
...
     @Override
-    public List<PatternWrapper> getPatterns() {
+    public Set<PatternWrapper> getPatterns() {
         final Set<PatternWrapper> patterns = new HashSet<>();
 
         if (isNewColumn() || !isNewFactPattern()) {
-            for (Pattern52 pattern52 : presenter.getModel().getPatterns()) {
-                patterns.add(new PatternWrapper(pattern52));
-            }
+            final BRLRuleModel brlRuleModel = new BRLRuleModel(presenter.getModel());
```
Comment (manstis) [inline drools-wb-screens/drools-wb-guided-dtable-editor/drools-wb-guided-dtable-editor-client/src/main/java/org/drools/workbench/screens/guided/dtable/client/wizard/column/plugins/ActionSetFactPlugin.java]: Delegate retrieval of bindings to ```BRLRuleModel``` (and not just ```Pattern52``` objects).

File: drools-wb-screens/drools-wb-guided-dtable-editor/drools-wb-guided-dtable-editor-client/src/main/java/org/drools/workbench/screens/guided/dtable/client/wizard/column/plugins/ActionWorkItemSetFieldPlugin.java [hunk 4]
```diff
@@ -409,25 +410,32 @@ public void setEntryPointName(final String entryPointName) {
...
     @Override
-    public List<PatternWrapper> getPatterns() {
+    public Set<PatternWrapper> getPatterns() {
         final Set<PatternWrapper> patterns = new HashSet<>();
 
         if (isNewColumn() || !isNewFactPattern()) {
-            for (Pattern52 pattern52 : presenter.getModel().getPatterns()) {
-                patterns.add(new PatternWrapper(pattern52));
-            }
+            final BRLRuleModel brlRuleModel = new BRLRuleModel(presenter.getModel());
```
Comment (manstis) [inline drools-wb-screens/drools-wb-guided-dtable-editor/drools-wb-guided-dtable-editor-client/src/main/java/org/drools/workbench/screens/guided/dtable/client/wizard/column/plugins/ActionWorkItemSetFieldPlugin.java]: Delegate retrieval of bindings to ```BRLRuleModel``` (and not just ```Pattern52``` objects).

File: drools-wb-screens/drools-wb-guided-dtable-editor/drools-wb-guided-dtable-editor-client/src/main/java/org/drools/workbench/screens/guided/dtable/client/wizard/column/plugins/ConditionColumnPlugin.java [hunk 5]
```diff
@@ -260,14 +262,19 @@ public void setEntryPointName(final String entryPointName) {
...
-    public List<PatternWrapper> getPatterns() {
+    public Set<PatternWrapper> getPatterns() {
         final Set<PatternWrapper> patterns = new HashSet<>();
-
-        for (Pattern52 pattern52 : model().getPatterns()) {
-            patterns.add(new PatternWrapper(pattern52));
-        }
-
-        return new ArrayList<>(patterns);
+        final BRLRuleModel brlRuleModel = new BRLRuleModel(presenter.getModel());
```
Comment (manstis) [inline drools-wb-screens/drools-wb-guided-dtable-editor/drools-wb-guided-dtable-editor-client/src/main/java/org/drools/workbench/screens/guided/dtable/client/wizard/column/plugins/ConditionColumnPlugin.java]: Delegate retrieval of bindings to ```BRLRuleModel``` (and not just ```Pattern52``` objects).

File: drools-wb-screens/drools-wb-guided-dtable-editor/drools-wb-guided-dtable-editor-client/src/main/java/org/drools/workbench/screens/guided/dtable/client/wizard/column/plugins/ActionWorkItemSetFieldPlugin.java [hunk 6]
```diff
@@ -500,11 +508,11 @@ public boolean isUpdateEngine() {
     }
 
     boolean isNewFactPattern() {
-        return !presenter
-                .getModel()
-                .getPatterns()
+        final BRLRuleModel brlRuleModel = new BRLRuleModel(presenter.getModel());
```
Comment (manstis) [inline drools-wb-screens/drools-wb-guided-dtable-editor/drools-wb-guided-dtable-editor-client/src/main/java/org/drools/workbench/screens/guided/dtable/client/wizard/column/plugins/ActionWorkItemSetFieldPlugin.java]: Delegate check to ```BRLRuleModel``` (and not just ```Pattern52``` objects).

File: drools-wb-screens/drools-wb-guided-dtable-editor/drools-wb-guided-dtable-editor-client/src/main/java/org/drools/workbench/screens/guided/dtable/client/wizard/column/plugins/commons/PatternWrapper.java [hunk 7]
```diff
@@ -127,4 +120,40 @@ public String getFactType() {
     public String getBoundName() {
         return boundName;
     }
+
+    @Override
+    public boolean equals(Object o) {
```
Comment (manstis) [inline drools-wb-screens/drools-wb-guided-dtable-editor/drools-wb-guided-dtable-editor-client/src/main/java/org/drools/workbench/screens/guided/dtable/client/wizard/column/plugins/commons/PatternWrapper.java]: Added ```equals( o )``` and ```hashCode()``` (generated by IDE, with addition of ```~~``` in ```hashCode()```) as we now use a ```Set``` for ```PatternWrapper```.

**Comment by manstis** [comment]
@jomarko The fix for https://issues.jboss.org/browse/GUVNOR-3409 really needs the issues around https://issues.jboss.org/browse/GUVNOR-3547 resolved first.. hence this PR first.. GUVNOR-3409 is more complicated and needs a little more thought but should follow this week.

**Comment by jomarko** [comment]
jenkins execute full downstream build

**Comment by manstis** [comment]
@karreiro Thanks.. updated.

**Comment by jomarko** [comment]
jenkins execute full downstream build

**Comment by manstis** [comment]
@karreiro Please could you review the changes.

**Pull request merged**